File: next-fx/src/test/java/com/rememberber/mootool/nextfx/domain/encode/EncodeEngineTest.java
package com.rememberber.mootool.nextfx.domain.encode;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class EncodeEngineTest {

    @Test
    void roundTripsUnicodeAndUtf8Hex() {
        assertThat(EncodeEngine.fromUnicode(EncodeEngine.toUnicode("Moo 工具 🚀"))).isEqualTo("Moo 工具 🚀");
        assertThat(EncodeEngine.hexToText(EncodeEngine.textToHex("你好 Moo"))).isEqualTo("你好 Moo");
        assertThat(EncodeEngine.toUnicode("Moo 工具 🚀")).contains("\\u5de5").contains("\\ud83d").contains("\\ude80");
    }

    @Test
    void roundTripsUrlTextInUtf8AndGb2312() {
        assertThat(EncodeEngine.urlDecode(
                EncodeEngine.urlEncode("你好 a/b", EncodeEngine.UrlCharset.UTF_8),
                EncodeEngine.UrlCharset.UTF_8
        )).isEqualTo("你好 a/b");
        assertThat(EncodeEngine.urlEncode("你好 a/b", EncodeEngine.UrlCharset.UTF_8))
                .isEqualTo("%E4%BD%A0%E5%A5%BD%20a%2Fb");
        assertThat(EncodeEngine.urlDecode(
                EncodeEngine.urlEncode("编码测试", EncodeEngine.UrlCharset.GB2312),
                EncodeEngine.UrlCharset.GB2312
        )).isEqualTo("编码测试");
    }

    @Test
    void convertsUnicodeCodePointsToAndFromDecimalOrHexAsciiLists() {
        assertThat(EncodeEngine.textToAscii("A中", EncodeEngine.AsciiFormat.DECIMAL)).isEqualTo("65 20013");
        assertThat(EncodeEngine.textToAscii("A中", EncodeEngine.AsciiFormat.HEX)).isEqualTo("41 4E2D");
        assertThat(EncodeEngine.asciiToText("65 4E2D")).isEqualTo("A中");
        assertThat(EncodeEngine.convert(
                EncodeEngine.Tab.ASCII,
                EncodeEngine.Direction.REVERSE,
                "65,0x4E2D",
                EncodeEngine.UrlCharset.UTF_8,
                EncodeEngine.AsciiFormat.DECIMAL
        )).isEqualTo("A中");
    }

    @Test
    void rejectsInvalidHexAndCodePoints() {
        assertThatThrownBy(() -> EncodeEngine.hexToText("zzz"))
                .isInstanceOf(EncodeException.class)
                .satisfies(error -> assertThat(((EncodeException) error).messageKey()).isEqualTo("encode.error.invalidHex"));
        assertThatThrownBy(() -> EncodeEngine.asciiToText("1114112"))
                .isInstanceOf(EncodeException.class)
                .satisfies(error -> assertThat(((EncodeException) error).messageKey()).isEqualTo("encode.error.invalidCodePoint"));
    }
}

File: next-compose/composeApp/src/desktopTest/kotlin/com/rememberber/mootool/next/compose/domain/EncodeEngineTest.kt
package com.rememberber.mootool.next.compose.domain

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

class EncodeEngineTest {
    @Test
    fun roundTripsUnicodeAndUtf8Hex() {
        assertEquals("Moo 工具 🚀", EncodeEngine.fromUnicode(EncodeEngine.toUnicode("Moo 工具 🚀")))
        assertEquals("你好 Moo", EncodeEngine.hexToText(EncodeEngine.textToHex("你好 Moo")))
        assertTrue(EncodeEngine.toUnicode("🚀").contains("\\u"))
    }

    @Test
    fun roundTripsUrlTextInUtf8AndGb2312() {
        assertEquals("你好 a/b", EncodeEngine.urlDecode(EncodeEngine.urlEncode("你好 a/b", UrlCharset.Utf8), UrlCharset.Utf8))
        assertEquals("编码测试", EncodeEngine.urlDecode(EncodeEngine.urlEncode("编码测试", UrlCharset.Gb2312), UrlCharset.Gb2312))
        assertTrue(EncodeEngine.urlEncode("a b", UrlCharset.Utf8).contains("%20"))
        assertEquals("a b", EncodeEngine.urlDecode("a+b", UrlCharset.Utf8))
    }

    @Test
    fun convertsCodePointsToDecimalOrHexLists() {
        assertEquals("65 20013", EncodeEngine.textToAscii("A中", AsciiFormat.Decimal))
        assertEquals("41 4E2D", EncodeEngine.textToAscii("A中", AsciiFormat.Hex))
        assertEquals("A中", EncodeEngine.asciiToText("65 4E2D"))
        assertEquals("", EncodeEngine.asciiToText("  "))
    }

    @Test
    fun rejectsBrokenHexTruncationAndUnmappableGb2312() {
        assertFailsWith<EncodeException> { EncodeEngine.hexToText("zzz") }
        assertFailsWith<EncodeException> { EncodeEngine.hexToText("abc") }
        assertFailsWith<EncodeException> { EncodeEngine.hexToText("ff") }
        assertFailsWith<EncodeException> { EncodeEngine.urlEncode("🚀", UrlCharset.Gb2312) }
        assertFailsWith<EncodeException> { EncodeEngine.asciiToText("1114112") }
    }

    @Test
    fun preservesNewlinesAndPercentInUrlRoundTrip() {
        val source = "line1\nline2 % path"
        assertEquals(source, EncodeEngine.urlDecode(EncodeEngine.urlEncode(source, UrlCharset.Utf8), UrlCharset.Utf8))
        assertEquals("MooTool 编码转换", EncodeEngine.fromUnicode(EncodeEngine.toUnicode("MooTool 编码转换")))
    }
}
